### PR TITLE
DOC: Adds Iñigo's week 6 and week 7 blogposts

### PR DESCRIPTION
**Pull request by itellaetxe**

### CODE REVIEW

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+This week was all about the Variational AutoEncoder. My mentors advised me to drop the TensorFlow implementation of the regression VAE I found last week, to instead directly integrate the variational and conditional characteristics in my AE implementation, following a more modular approach. This was a good decision, as adapting third party code to one's needs is often a bit of a mess (it already started being a mess, so yeah). Also, once the variational part is done, implementing the conditional should not be that hard.
+
+To provide a bit of intuition behind the VAE, let me illustrate first the "vanilla" AE. This is a neural network that compresses the input data into a reduced representation, and then tries to reconstruct the original data from this representation. We refer to "latent space" to the place where the compressed data representations live. So, once the AE is good at compressing data to the latent space, we could take a sample from it and generate new data.
+The objective of the AE is to minimize the difference between the input data and the generated data, also called the "reconstruction loss".
+
+.. image:: /_static/images/inigo_vanilla_autoencoder.png
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:20]: Did you create yourself these figures? Which is their license? If we do not know that, we may not be entitle to use them?

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+==================================================================================
+
+
+.. post:: July 06 2024
+    :author: Iñigo Tellaetxe
+    :tags: google
+    :category: gsoc
+
+
+
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:11]: No need to add unnecessary blank lines besides one.

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 3]
```diff
@@ -0,0 +1,48 @@
+Week 6 into GSoC 2024: Stuck with the Variational AutoEncoder, problems with Keras
+==================================================================================
+
+
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:4]: No need to add an additional blank line.

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+What I did this week
+~~~~~~~~~~~~~~~~~~~~
+
+This week was all about the Variational AutoEncoder. My mentors advised me to drop the TensorFlow implementation of the regression VAE I found last week, to instead directly integrate the variational and conditional characteristics in my AE implementation, following a more modular approach. This was a good decision, as adapting third party code to one's needs is often a bit of a mess (it already started being a mess, so yeah). Also, once the variational part is done, implementing the conditional should not be that hard.
+
+To provide a bit of intuition behind the VAE, let me illustrate first the "vanilla" AE. This is a neural network that compresses the input data into a reduced representation, and then tries to reconstruct the original data from this representation. We refer to "latent space" to the place where the compressed data representations live. So, once the AE is good at compressing data to the latent space, we could take a sample from it and generate new data.
+The objective of the AE is to minimize the difference between the input data and the generated data, also called the "reconstruction loss".
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:18]: Add a blank line to start a new paragraph.

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+.. image:: /_static/images/inigo_vanilla_autoencoder.png
+    :alt: Vanilla AutoEncoder diagram.
+    :width: 600
+
+
+Usually the problem with AEs is that the latent space is full of "holes", i.e. it is a set of points, meaning that if a sample is taken between two already encoded samples, the generated data will not be an interpolation between the two samples, but a random sample. To solve this, the VAE seeks to "regularize" the latent space, encoding/compressing the input data into distributions that live in it, instead of single points. This way, the latent space is continuous and the interpolation between two samples is meaningful and more likely to follow the data distribution.
+The VAE does this by adding a "regularization loss" to the AE, which is the Kullback-Leibler divergence between the latent space distribution and a prior distribution (usually a normal distribution). If you are curious, you can find a nice explanation about the differences between AEs and VAEs `here <https://towardsdatascience.com/understanding-variational-autoencoders-vaes-f70510919f73>`_.
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:26]: Add a blank line to start a new paragraph.

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+.. image:: /_static/images/inigo_variational_autoencoder.png
+    :alt: Variational AutoEncoder diagram.
+    :width: 600
+
+
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:32]: Same comment as above about blank lines.

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+.. image:: /_static/images/inigo_vanilla_autoencoder.png
+    :alt: Vanilla AutoEncoder diagram.
+    :width: 600
+
+
+Usually the problem with AEs is that the latent space is full of "holes", i.e. it is a set of points, meaning that if a sample is taken between two already encoded samples, the generated data will not be an interpolation between the two samples, but a random sample. To solve this, the VAE seeks to "regularize" the latent space, encoding/compressing the input data into distributions that live in it, instead of single points. This way, the latent space is continuous and the interpolation between two samples is meaningful and more likely to follow the data distribution.
+The VAE does this by adding a "regularization loss" to the AE, which is the Kullback-Leibler divergence between the latent space distribution and a prior distribution (usually a normal distribution). If you are curious, you can find a nice explanation about the differences between AEs and VAEs `here <https://towardsdatascience.com/understanding-variational-autoencoders-vaes-f70510919f73>`_.
+
+.. image:: /_static/images/inigo_variational_autoencoder.png
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:28]: Same comment as above about the figure.

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    :alt: Variational AutoEncoder diagram.
+    :width: 600
+
+
+Thus, I started by implementing the VAE in Keras, and I must say that it was not as easy as I thought it would be. I used `this VAE example from Keras <https://keras.io/examples/generative/vae/>`_, adding a ``ReparametrizationTrickSampling`` layer between the ``Encoder`` and the ``Decoder``.
+The main problem was that the Keras implementation was not behaving as expected because it was giving me problems of model initialization, so I encapsulated the ``Encoder`` and the ``Decoder`` parts into individual ``keras.Model`` class instances, to put them all together under a wrapping Model. Doing this was a bit painful, but it worked.
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:34]: If new paragraph, add a blank line.

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+The problem was that I was not being able to train the model because the loss was constantly ``nan`` values.
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:35]: If new paragraph, add a blank line.

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+What is coming up next week
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Next week I must correct the ``nan`` problem when training because I am not sure what is causing it. I am using exponential operations in the KL loss computation and in the ``ReparametrizationTrickSampling`` layer, what could explode if the values of the exponent are too big. I will explore this.
+
+
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:42]: Same comment as above about blank lines.

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+Thus, I started by implementing the VAE in Keras, and I must say that it was not as easy as I thought it would be. I used `this VAE example from Keras <https://keras.io/examples/generative/vae/>`_, adding a ``ReparametrizationTrickSampling`` layer between the ``Encoder`` and the ``Decoder``.
+The main problem was that the Keras implementation was not behaving as expected because it was giving me problems of model initialization, so I encapsulated the ``Encoder`` and the ``Decoder`` parts into individual ``keras.Model`` class instances, to put them all together under a wrapping Model. Doing this was a bit painful, but it worked.
+The problem was that I was not being able to train the model because the loss was constantly ``nan`` values.
+
+What is coming up next week
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Next week I must correct the ``nan`` problem when training because I am not sure what is causing it. I am using exponential operations in the KL loss computation and in the ``ReparametrizationTrickSampling`` layer, what could explode if the values of the exponent are too big. I will explore this.
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:40]: (...) layer, which could return excessively large values if the exponent value is too big, leading to an exploding gradients issue.

File: posts/2024/2024_07_12_Inigo_week_7.rst [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+======================================================================
+
+
+.. post:: July 12 2024
+    :author: Iñigo Tellaetxe
+    :tags: google
+    :category: gsoc
+
+
+
```
Comment (jhlegarreta) [inline posts/2024/2024_07_12_Inigo_week_7.rst:11]: Same comment as above about blank lines.

File: posts/2024/2024_07_12_Inigo_week_7.rst [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+What I did this week
+~~~~~~~~~~~~~~~~~~~~
+
+Finally, I figured out how to solve the ``nan`` value problem in the VAE training. As I suspected, the values that the ``ReparametrizationTrickSampling`` layer was getting were too big for the exponential operations. I use the exponential operation because I am treating the Encoder output as the log variance of the latent space distribution, and for sampling we need the standard deviation. We use the log variance instead of the standard deviation for avoiding computing logarithms.
+The solution consisted of two measures: first, clipping the values of the exponential operations to the ``[-1e10, 1e10]`` range. Then, adding batch normalization layers after each convolution in the Encoder enforcing mean=0 and variance=1, avoids the values from getting too big. I did not use batch normalization in the fully connected layers that output the mean and the log variance of the latent distribution because I did not want to constrain the latent space too much with distributions with the mentioned characteristics. These layers should capture other characteristics of the data distribution.
```
Comment (jhlegarreta) [inline posts/2024/2024_07_12_Inigo_week_7.rst:16]: If new paragraph, blank line between them.

Comment (jhlegarreta) [inline posts/2024/2024_07_12_Inigo_week_7.rst:16]: `(...) range. Then, adding batch normalization layers after each convolution in the Encoder enforcing mean=0 and variance=1, avoids the values from getting too big.`

`(...) range; second, adding batch normalization layers after each convolution in the Encoder enforces a 0 mean and unit variance, preventing large values at the output.`

File: posts/2024/2024_07_12_Inigo_week_7.rst [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+You can see a preliminary result of training the Variational AutoEncoder below, with the exact same training parameters as the AutoEncoder, but only for 50 epochs for proving the concept.
+
+.. image:: /_static/images/inigo_preliminary_vae_result_fibercup.png
+    :alt: Preliminary reconstruction result after training the VAE for 50 epochs with the FiberCup dataset.
+    :width: 600
+
+
```
Comment (jhlegarreta) [inline posts/2024/2024_07_12_Inigo_week_7.rst:23]: Same comment as above about blank lines.

File: posts/2024/2024_07_12_Inigo_week_7.rst [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+I should mention that the variance of the 0-mean distribution from which ``epsilon`` is sampled in the ``ReparametrizationTrickSampling`` layer can be adjusted as a hyperparameter. I set it to 1 for now, but I could explore this hyperparameter in the future, but first I need to learn about the theoretical implications of modifying it. 
+
+On the other hand, I started a discussion in a `PR <https://github.com/itellaetxe/tractoencoder_gsoc/pull/1>`_ that includes all these changes to make it easier for my mentors and my GSoC colleagues to review my code. This change was suggested by my mentors to make my advances more public and accessible to feedback.
+
+
```
Comment (jhlegarreta) [inline posts/2024/2024_07_12_Inigo_week_7.rst:28]: Same comment as above about blank lines.

File: posts/2024/2024_07_06_Inigo_week_6.rst [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+The objective of the AE is to minimize the difference between the input data and the generated data, also called the "reconstruction loss".
+
+.. image:: /_static/images/inigo_vanilla_autoencoder.png
+    :alt: Vanilla AutoEncoder diagram.
+    :width: 600
+
+
+Usually the problem with AEs is that the latent space is full of "holes", i.e. it is a set of points, meaning that if a sample is taken between two already encoded samples, the generated data will not be an interpolation between the two samples, but a random sample. To solve this, the VAE seeks to "regularize" the latent space, encoding/compressing the input data into distributions that live in it, instead of single points. This way, the latent space is continuous and the interpolation between two samples is meaningful and more likely to follow the data distribution.
```
Comment (jhlegarreta) [inline posts/2024/2024_07_06_Inigo_week_6.rst:25]: No need to change this, but bear in mind that as you say, the latent space that we will get will have a normal distribution, across all bundles: so one single normal across all bundles. Some time ago I think I shared a paper where they showed the type of tractography reconstructions that they got using AEs, VAEs, VQ-AEs, etc.

**Review by jhlegarreta — COMMENTED** [review]
Style comments mostly.

Thanks Iñigo.

File: posts/2024/2024_07_06_Inigo_week_6.rst (hunk 1 above)
Comment (itellaetxe) [reply posts/2024/2024_07_06_Inigo_week_6.rst:20]: Yes they are mine! I used [draw.io](https://app.diagrams.net/) to create them

**Review by robinroy03 — APPROVED** [review]
Good work @itellaetxe!
I like the week-6 diagrams :)

Approving.

File: posts/2024/2024_07_06_Inigo_week_6.rst (hunk 1 above)
Comment (jhlegarreta) [reply posts/2024/2024_07_06_Inigo_week_6.rst:18]: Thanks for having worked on the diagrams.

**Review by pjsjongsung — APPROVED** [review]

**Review by deka27 — APPROVED** [review]

**Review by skoudoro — APPROVED** [review]
Thank you, it is all clear. Merging

**Comment by itellaetxe** [comment]
@skoudoro I just posted the content in Mastodon linking to the deployed dipy.org.
I am currently writing the blogs of weeks 8 and 9 from the notes I took back then. I will open PR soon.
